packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -278,6 +341,151 @@ class TextEditingController extends ValueNotifier<TextEditingValue> {
   }
 }
 
+/// A [TextEditingController] that contains a list of [TextEditingInlineSpanReplacement]s that
+/// insert custom [InlineSpan]s in place of matched [Pattern]s.
+///
+/// This controller should be passed [TextEditingInlineSpanReplacement], each of which contains a
+/// a [Pattern] to match with and a generator function to generate an [InlineSpan] to replace
+/// the matched [Pattern] with based on the matched string.
+///
+/// See [TextEditingInlineSpanReplacement] for example replacements to provide this class with.
+class ReplacementTextEditingController extends TextEditingController {
+
+  /// Constructs a controller with optional text that handles the provided list of replacements.
+  ReplacementTextEditingController({
+    String? text,
+    required this.replacements,
+    this.composingRegionPrioritized = false,
+  }) : assert(replacements.isNotEmpty), super(text: text);
+
+  /// Creates a controller for an editable text field from an initial [TextEditingValue].
+  ///
+  /// This constructor treats a null [value] argument as if it were
+  /// [TextEditingValue.empty].
+  ReplacementTextEditingController.fromValue(TextEditingValue? value, {
+    required this.replacements,
+    this.composingRegionPrioritized = false
+  }) : assert(replacements.isNotEmpty), super.fromValue(value);
+
+  /// The [TextEditingInlineSpanReplacement]s that are evaluated on the editing value.
+  ///
+  /// Each replacement is evaluated in order from first to last. If multiple replacement
+  /// [Pattern]s match against the same range of text,
+  /// the first replacement will be used and any additional matches that overlap will
+  /// be ignored.
+  ///
+  /// For example, if given replacements with patterns of '{hello}' and
+  /// 'hello', only the first replacement will be used as the second one is always
+  /// overlapping with the first.git@github.com:flutter/packages.git
+  final List<TextEditingInlineSpanReplacement> replacements;
+
+  /// If composing regions should be matched against for replacements.
+  ///
+  /// When true, composing regions are added before any patterns are applied,
+  /// invalidating the composing region from being matched against.
+  ///
+  /// When false, composing regions are added after all patterns are matched
+  /// and replacements made. This means that composing region may sometimes
+  /// fail to display if the composing region matches against of the the
+  /// replacement patterns.
+  final bool composingRegionPrioritized;
```

Comment:
why "prioritized"? Isn't this basically saying whether the composingRegion should be ignored for replacements? E.g. when this is true, a long as a string is part of the composing region, it cannot be replaced with a replacement?

Reply:
Renamed to `composingRegionReplaceable`